### DISCUSSION TITLE
Remove Category from Code Scanning

### DISCUSSION THREAD

**Discussion by TWiStErRob** [community/community#37188] [Code Security] [answered]
Hi, I'm looking to remove this highlighted line:
![image](https://user-images.githubusercontent.com/2906988/197721312-c76b18a3-ced8-46e9-9fc9-f429ba56fab4.png)

https://github.com/TWiStErRob/net.twisterrob.cinema/pull/252/checks?check_run_id=9084675104

History:
 * I've set up CI to run Detekt in one job, upload the sarif files, and then in a separate job publish it.
 * Then I merged these two jobs into one.

Because of this move the first line `publish-detekt-results-gradle` is outdated and should be removed. I even get warnings about it:
> Warning: Code scanning cannot determine the alerts introduced or fixed by this pull request, because 1 analysis was not found.

![image](https://user-images.githubusercontent.com/2906988/197723972-4caad4cd-f7b4-4271-8460-36a0104e1ade.png)

But my only recourse I can see is if I start double-publishing the same results with different categories. I'm looking for a way to delete this no-unused category.

**Comment by TWiStErRob** [comment]
Note: I looked in Settings, but the options are pretty lean:
![image](https://user-images.githubusercontent.com/2906988/197724389-56c82067-9481-4dc3-b26a-324f1bf9181a.png)

**Comment by starcke** [comment, marked as answer]
It is possible to delete old analyses (like the one you see here for `publish-detekt-results-gradle`) using the REST API:
https://docs.github.com/en/rest/code-scanning#delete-a-code-scanning-analysis-from-a-repository

In order to get the analysis IDs to delete, you need to look through the list analyses REST API:
https://docs.github.com/en/rest/code-scanning#list-code-scanning-analyses-for-a-repository

For each analysis you should look at whether the `category` field contains `publish-detekt-results-gradle` and whether they are `deletable: true`, and if so delete it (possibly specifying `confirm_delete`).

**Reply by TWiStErRob** [reply]
Amazing idea! Thanks for the pointers.

I started experimenting with `gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses?ref=refs/heads/master --paginate` and managed to see the offending category.

Then I used the wizardry on https://cli.github.com/manual/gh_help_formatting to build this (triple-quote escape for Windows `cmd`, ymmv):
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses?ref=refs/heads/master --paginate --jq ".[] | select(.category == """.github/workflows/Heroku.yml:publish-detekt-results-gradle""") | { id: .id, deletable: .deletable }"
```

There was only one deletable entry, I was able to see the details:
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses/33491792
```
```json
{
  "ref": "refs/heads/master",
  "commit_sha": "45f5b1682eed60d3a02debf1810a39bce148cc90",
  "analysis_key": ".github/workflows/Heroku.yml:publish-detekt-results-gradle",
  "environment": "{}",
  "category": ".github/workflows/Heroku.yml:publish-detekt-results-gradle",
  "error": "",
  "created_at": "2022-06-24T20:08:09Z",
  "results_count": 0,
  "rules_count": 191,
  "id": 33491792,
  "url": "https://api.github.com/repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses/33491792",
  "sarif_id": "5e316016-f3f9-11ec-986b-de4b3ee99abf",
  "tool": {
    "name": "detekt",
    "guid": "022ca8c2-f6a2-4c95-b107-bb72c43263f3",
    "version": "1.20.0"
  },
  "deletable": true,
  "warning": ""
}
```
This actually matched the problematic one in the screenshot:
> ✔️ Found for 45f5b16

So I deleted it (as you expected `confirm_delete` was required):
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses/33491792 --method DELETE
```
```json
{
  "message": "Analysis is last of its type and deletion may result in the loss of historical alert data. Please specify confirm_delete.",
  "documentation_url": "https://docs.github.com/rest/reference/code-scanning#delete-a-code-scanning-analysis-from-a-repository"
}
gh: Analysis is last of its type and deletion may result in the loss of historical alert data. Please specify confirm_delete. (HTTP 400)
gh: This API operation needs the "admin:repo_hook" scope. To request it, run:  gh auth refresh -h github.com -s admin:repo_hook
```
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses/33491792?confirm_delete=true --method DELETE
```
```json
{
  "next_analysis_url": null,
  "confirm_delete_url": null
}
```

I re-queried using the filter and there were no other `deletable` ones:
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses?ref=refs/heads/master --paginate --jq ".[] | select(.category == """.github/workflows/Heroku.yml:publish-detekt-results-gradle""") | select(.deletable) | .id"
```

I reran the CI but I still observed the same problem, but now with a different commit hash. So I'll need to go through the loop-d-loop described in https://docs.github.com/en/rest/code-scanning#delete-a-code-scanning-analysis-from-a-repository

**Reply by TWiStErRob** [reply]
Loop with `gh api` CLI (at this point I removed the branch filter and just nuking everything for this category):
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses --paginate --jq ".[] | select(.deletable and .category == """.github/workflows/Heroku.yml:publish-detekt-results-gradle""") | """gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses/""" + (.id | tostring) + """?confirm_delete=true --method DELETE""""
```
which when executed makes all of them returned:
```json
{
  "next_analysis_url": null,
  "confirm_delete_url": null
}
```
which is weird, because now the list contains no `deletable` entries:
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses --paginate --jq ".[] | select(.category == """.github/workflows/Heroku.yml:publish-detekt-results-gradle""") | { id: .id, deletable: .deletable, ref: .ref, tool: .tool.guid }"
```
I tried removing the `jq` filter, just dumping everything:
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses --paginate --jq=".[] | [(.id | tostring), .deletable, .tool.guid, .ref, .category] | join(""",""")" > all.csv
```
but there are still no deletable entries. ⛔

⚠ I reported this empty list at https://github.com/orgs/community/discussions/37897

**Reply by starcke** [reply]
> I reran the CI but I still observed the same problem, but now with a different commit hash. So I'll need to go through the loop-d-loop described in https://docs.github.com/en/rest/code-scanning#delete-a-code-scanning-analysis-from-a-repository

> :warning: I reported this empty list at https://github.com/orgs/community/discussions/37897

I think what you have done is correct and the output matches what I would expect. The reason is that the delete allows you to delete the most recent analysis (and replace it with the second most recent analysis) - kind of like an _undo_ operation. In this case you have undone all you could, and the alerts from the old config should not show up anymore.

So I think there is another reason for why the PRs still show the old category. It could be because the PR is based on an old master commit (`45f5b1682eed60d3a02debf1810a39bce148cc90`) back from when the `publish-detekt-results-gradle` category was still in the file: https://github.com/TWiStErRob/net.twisterrob.cinema/blob/45f5b1682eed60d3a02debf1810a39bce148cc90/.github/workflows/Heroku.yml.

I couldnt really see if that was the case as the PR was squashed when merged, but to test this you could try opening a fresh PR from current master and see if indeed the `publish-detekt-results-gradle` category is gone.

**Reply by TWiStErRob** [reply]
> I think what you have done is correct and the output matches what I would expect.

Yes, the first part "I reran the CI but I still observed the same problem, but now with a different commit hash" was expected, but the empty list is not. According to documentation you can "undo" until it's all gone, but with no next `deletable` entry, you're stuck.

It is true that after deleting all I could I forgot to re-run the CI, because I didn't expect any change, and now days later, it's still the same problem. If you have a look at [a recent run created today based on HEAD of `master`](https://github.com/TWiStErRob/net.twisterrob.cinema/runs/9441016408) you can see that it's still finding the `master` copy for `publish-detekt-results-gradle` category:
![image](https://user-images.githubusercontent.com/2906988/201417583-08e7b145-4740-401b-9eec-51caf6646682.png)
because I didn't actually delete all of them.

https://github.com/orgs/community/discussions/37897 is the cause, and needs to be fixed because this answer is workable in real life.

Based on the screenshot above and this command's output:
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses --paginate --jq ".[] | select(.category == """.github/workflows/Heroku.yml:publish-detekt-results-gradle""") | { id: .id, deletable: .deletable, ref: .ref, tool: .tool.guid, sha: .commit_sha }"
```
```json
{"deletable":false,"id":33418631,"ref":"refs/heads/master","sha":"80c16ab5f087fc641145afcc7fbb162fa4804e45","tool":"022ca8c2-f6a2-4c95-b107-bb72c43263f3"}
```
should be `deletable:true` as it's the next head of undo.

I'm [trying (probably protected)](https://support.github.com/ticket/personal/0/1875650) to contact support about https://github.com/orgs/community/discussions/37897.

**Reply by starcke** [reply]
> It is true that after deleting all I could I forgot to re-run the CI, because I didn't expect any change, and now days later, it's still the same problem. If you have a look at [a recent run created today based on HEAD of master](https://github.com/TWiStErRob/net.twisterrob.cinema/runs/9441016408) you can see that it's still finding the master copy for publish-detekt-results-gradle category:

> https://github.com/orgs/community/discussions/37897 is the cause, and needs to be fixed because this answer is workable in real life.

I investigated this further after you confirmed it reproduced on a fresh PR and I think I found an issue with the way we try to match up the categories when they are not all there. Basically the delete you already did should have made it so that `publish-detekt-results-gradle` no longer shows up, but due to a bug (with the fetch and not the delete) it seems that in rare cases it might show up anyways.

I'll try to get this confirmed and fixed if indeed that is the case.

**Reply by starcke** [reply]
We did indeed find a bug and pushed a fix, please let me know if the issue persists after you rerun the checks.

**Reply by TWiStErRob** [reply]
Confirmed fixed, `Heroku.yml:publish-detekt-results-gradle` category is gone for recent PRs against `master`: https://github.com/TWiStErRob/net.twisterrob.cinema/pull/321/checks?check_run_id=10083552809

![image](https://user-images.githubusercontent.com/2906988/207621644-a54937c1-d50b-4600-a7fb-4119331cf87e.png)

---

It is however still very strange that this returns no deletable results at all:
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses --paginate --jq ".[] | select(.category == """.github/workflows/Heroku.yml:publish-detekt-results-gradle""") | { id: .id, deletable: .deletable, ref: .ref, tool: .tool.guid }"
```
because according to the [docs](https://docs.github.com/en/rest/code-scanning?apiVersion=2022-11-28#delete-a-code-scanning-analysis-from-a-repository):
> where a set is determined by a unique combination of analysis values: ref, tool, category

When execute the above I get a lot similar looking things. Here's a focused example:
```shell
gh api /repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses --paginate --jq ".[] | select(.category == """.github/workflows/Heroku.yml:publish-detekt-results-gradle""" and .ref == """refs/heads/kotlin16""")"
```
For the same ref, and same category all the tools are the same, yet they don't form a set where the head can be chopped off with a DELETE:
```json
[
	{
		"analysis_key": ".github/workflows/Heroku.yml:publish-detekt-results-gradle",
		"category": ".github/workflows/Heroku.yml:publish-detekt-results-gradle",
		"commit_sha": "0a822e5b9412a938967c6208796cc15e9caa3f9e",
		"created_at": "2022-04-03T11:17:30Z",
		"deletable": false,
		"environment": "{}",
		"error": "",
		"id": 25361394,
		"ref": "refs/heads/kotlin16",
		"results_count": 1,
		"rules_count": 186,
		"sarif_id": "a691add4-b33f-11ec-8998-6c9e713a3513",
		"tool": {
			"guid": "022ca8c2-f6a2-4c95-b107-bb72c43263f3",
			"name": "detekt",
			"version": "1.19.0"
		},
		"url": "https://api.github.com/repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses/25361394",
		"warning": ""
	},
	{
		"analysis_key": ".github/workflows/Heroku.yml:publish-detekt-results-gradle",
		"category": ".github/workflows/Heroku.yml:publish-detekt-results-gradle",
		"commit_sha": "25e23d5b839e81f33e8032a83a0fe5c840af6714",
		"created_at": "2022-04-03T10:25:54Z",
		"deletable": false,
		"environment": "{}",
		"error": "",
		"id": 25360088,
		"ref": "refs/heads/kotlin16",
		"results_count": 1,
		"rules_count": 186,
		"sarif_id": "71955a56-b338-11ec-99be-78c051e4cec8",
		"tool": {
			"guid": "022ca8c2-f6a2-4c95-b107-bb72c43263f3",
			"name": "detekt",
			"version": "1.19.0"
		},
		"url": "https://api.github.com/repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses/25360088",
		"warning": ""
	},
	{
		"analysis_key": ".github/workflows/Heroku.yml:publish-detekt-results-gradle",
		"category": ".github/workflows/Heroku.yml:publish-detekt-results-gradle",
		"commit_sha": "f8b88ddfff06b5546d9488a309a683f1432190be",
		"created_at": "2022-04-03T09:41:53Z",
		"deletable": false,
		"environment": "{}",
		"error": "",
		"id": 25359024,
		"ref": "refs/heads/kotlin16",
		"results_count": 0,
		"rules_count": 186,
		"sarif_id": "4a1281b2-b332-11ec-9f2c-72d978d7ab9c",
		"tool": {
			"guid": "022ca8c2-f6a2-4c95-b107-bb72c43263f3",
			"name": "detekt",
			"version": "1.19.0"
		},
		"url": "https://api.github.com/repos/TWiStErRob/net.twisterrob.cinema/code-scanning/analyses/25359024",
		"warning": ""
	}
]
```
I had a quick look at the [`kotlin16`](https://github.com/TWiStErRob/net.twisterrob.cinema/pull/103) branch and the 3 commits listed above are:
 * f8b88dd before force push
 * 25e23d5 the force push
 * 0a822e5 after force push

This is a deleted branch, so it doesn't bother that much, but there are a lot of these for `master` too.
